service/service.go

```diff
@@ -342,25 +359,35 @@ func (app *Application) execute() {
 
 	app.asyncErrorChannel = make(chan error)
 
-	// Setup everything.
+	app.v.OnConfigChange(func(e fsnotify.Event) {
+		app.asyncErrorChannel <- errReloadConfig
+	})
+
 	app.setupTelemetry(ballastSizeBytes)
+
+	err := errReloadConfig
+	for err == errReloadConfig {
+		app.setup()
+		err = app.runAndWaitForShutdownEvent()
+		app.shutdown(ballast)
+	}
+
+	AppTelemetry.shutdown()
+	app.logger.Info("Shutdown complete.")
+}
+
+func (app *Application) setup() {
 	app.setupConfigurationComponents()
 	app.notifyPipelineReady()
+}
 
-	// Everything is ready, now run until an event requiring shutdown happens.
-	app.runAndWaitForShutdownEvent()
-
-	// Begin shutdown sequence.
+func (app *Application) shutdown(ballast []byte) {
 	runtime.KeepAlive(ballast)
```

Comment:
I would stick this `runtime.KeepAlive(ballast)` call in the `execute` function right before `AppTelemetry.shutdown()`.

Reply:
ok then I don't have to pass it around.